### PR TITLE
handle dsr with teid 0 and detach state guard timeout handling

### PR DESCRIPTION
**Pull request by badhrinathpa**
handle dsr with teid 0 and detach state guard timeout handling

### CODE REVIEW

File: scripts/SMCodeGen/codeGen.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 from os.path import join
 import utils
 #set path where you have Template-Toolkit 
-#sys.path.append("/users/ajayonf/Nucleus.Ajay/scripts/SMCodeGen/Template-Toolkit-Python")
+sys.path.append("/Users/badhrinathpa/bads_change/nucleus/master/bad_master/Nucleus/scripts/Template-Toolkit-Python")
```
Comment (thakurajayL) [inline scripts/SMCodeGen/codeGen.py:11]: Opps...looks like I committed this ...May be worth to take this line out.

File: src/gtpV2Codec/gtpV2Stack.cpp [hunk 2]
```diff
@@ -45,6 +45,49 @@ GtpV2Stack::~GtpV2Stack ()
...
+ ******************************************/
+int GtpV2Stack::delSeqKey( uint32_t key )
+{
+    std::lock_guard<std::mutex> lock(seq_ueidx_map_mutex);
+
+    return seq_ue_idx_map.erase( key );
+}
+
+/******************************************
+ * Find cb with given mTmsi from mTmsi_cb_id_map
```
Comment (thakurajayL) [inline src/gtpV2Codec/gtpV2Stack.cpp:76]: comments need correction

File: src/s11/handlers/s11_DS_resp_handler.c [hunk 3]
```diff
@@ -42,7 +43,10 @@ s11_DS_resp_handler(MsgBuffer* message, GtpV2MessageHeader* hdr)
 	/*Check whether has teid flag is set.
 	 * Also check whether this check is needed for DSR.
 	 * */
-	dsr_info.ue_idx = hdr->teid;	
+	//dsr_info.ue_idx = hdr->teid;
```
Comment (thakurajayL) [inline src/s11/handlers/s11_DS_resp_handler.c:46]: to keep things safe..Can we use ue_idx only when hdr->Teid is 0 ? Though current changes look good...add log if hdr->Teid != 0  && (hdr->Teid != ue_index). 

This will happen rarely when there is cross connection of tunnels.  You may do it or skip this.

**Comment by yoooou** [comment]
retest this please

**Comment by yoooou** [comment]
retest this please

File: src/gtpV2Codec/gtpV2Stack.cpp [hunk 4]
```diff
@@ -45,6 +45,48 @@ GtpV2Stack::~GtpV2Stack ()
     // TODO Auto-generated destructor stub
 }
 
+/******************************************
+ * Add Seq as key and ue_idx as value to map
+ ******************************************/
```
Comment (asreek) [inline src/gtpV2Codec/gtpV2Stack.cpp:50]: GtpV2Stack is an auto generated file. So next time we run the script these changes will be lost. We should move this to a different file

**Comment by yoooou** [comment]
retest this please

**Comment by yoooou** [comment]
retest this please

**Comment by badhrinathpa** [comment]
retest this please

File: include/s11/s11_options.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2019-present Open Networking Foundation
```
Comment (thakurajayL) [inline include/s11/s11_options.h:2]: is this file required ?

File: src/gtpV2Codec/gtpV2Stack.h [hunk 6]
```diff
@@ -55,6 +57,9 @@ class GtpV2Stack {
     bool decodeMessage(GtpV2MessageHeader& msgHeader, MsgBuffer& buffer,
                  void* data_p = NULL);
     void display_v(Uint8 msgType, Debug& stream, void* data_p = NULL);
+
+	std::map<uint32_t,uint32_t> seq_ue_idx_map;
```
Comment (thakurajayL) [inline src/gtpV2Codec/gtpV2Stack.h:61]: I thought we are not using this ?

File: include/s11/s11_options.h (hunk 5 above)
Comment (badhrinathpa) [reply include/s11/s11_options.h:2]: s11 was currently using options file from s1ap. this file removes the dependency between s1ap and s11 and can be used for common functions needed in s11.

File: src/gtpV2Codec/gtpV2Stack.h (hunk 6 above)
Comment (badhrinathpa) [reply src/gtpV2Codec/gtpV2Stack.h:61]: will remove.

File: src/mme-app/actionHandlers/ueInitDetachActionHandlers.cpp [hunk 7]
```diff
@@ -52,6 +52,7 @@ ActStatus ActionHandlers::del_session_req(SM::ControlBlock& cb)
 	
 	struct DS_Q_msg g_ds_msg;
 	g_ds_msg.msg_type = delete_session_request;
+	g_ds_msg.ue_idx = ue_ctxt->getContextID();
```
Comment (thakurajayL) [inline src/mme-app/actionHandlers/ueInitDetachActionHandlers.cpp:55]: No sgw address/teid ?

File: src/mme-app/structs.cpp [hunk 8]
```diff
@@ -490,21 +491,11 @@ bool DigitRegister15::operator == ( const DigitRegister15& data_i )const
...
-			if(digitsArray[i] > data_i.digitsArray[i])
-			{
-				rc = false;
-				break;
-			}
-		}
-	}
-	return rc;
+    bool rc = true;
+    int data_cmp = memcmp(digitsArray, data_i.digitsArray,  sizeof( data_i.digitsArray));
```
Comment (thakurajayL) [inline src/mme-app/structs.cpp:495]: Does this work ?

Comment (badhrinathpa) [reply src/mme-app/structs.cpp:495]: yes. tested locally.

File: src/mme-app/actionHandlers/ueInitDetachActionHandlers.cpp (hunk 7 above)
Comment (badhrinathpa) [reply src/mme-app/actionHandlers/ueInitDetachActionHandlers.cpp:55]: we send s11 fteid as another parameter.
